### PR TITLE
lexer: add fstring tokens pep 701

### PR DESCRIPTION
**Pull request by Glyphack**
Please red the following section:
https://peps.python.org/pep-0701/#new-tokens

### CODE REVIEW

File: src/lexer/diagnostics.rs [hunk 1]
```diff
@@ -2,6 +2,6 @@ use miette::{self, Diagnostic};
 use thiserror::{self, Error};
 
 #[derive(Debug, Error, Diagnostic)]
-#[error("Syntax error")]
+#[error("Syntax error {0}")]
```
Comment (frustak) [inline src/lexer/diagnostics.rs:5]: What does this change do?

File: src/lexer/lexer.rs [hunk 2]
```diff
@@ -17,6 +17,10 @@ pub struct Lexer {
     /// because the first line is always at indentation level 0
     indent_stack: Vec<usize>,
     nesting: i8,
+    // This stack means we are in a fstring that started with
+    // character at the top of the stack
+    fstring_stack: Vec<String>,
+    inside_fstring_bracket: i8,
```
Comment (frustak) [inline src/lexer/lexer.rs:23]: Can you add a comment about `inside_fstring_bracket`? I don't understand what it is.

File: src/lexer/lexer.rs [hunk 3]
```diff
@@ -53,19 +60,100 @@ impl Lexer {
...
+        let start_of_line = self.start_of_line;
+        let inside_fstring_bracket = self.inside_fstring_bracket;
         let token = self.next_token();
         self.current = current;
         self.nesting = nesting;
+        self.fstring_stack = fstring_stack;
+        self.start_of_line = start_of_line;
+        self.inside_fstring_bracket = inside_fstring_bracket;
         token
     }
```
Comment (frustak) [inline src/lexer/lexer.rs:73]: I'm a bit confused about `peek_token`. why does it mutate `self`, Maybe we should call it something other than `peek`?

File: src/lexer/lexer.rs [hunk 4]
```diff
@@ -53,19 +60,100 @@ impl Lexer {
...
+            if self.peek() == Some('}') {
+                self.next();
+                self.inside_fstring_bracket -= 1;
+                return Some(Kind::RightBracket);
+            }
+            // if we are inside a bracket return none
+            // and let the other tokens be matched
+            return None;
+        }
+        println!("not inside the bracket");
```
Comment (frustak) [inline src/lexer/lexer.rs:86]: Yes

File: src/lexer/lexer.rs [hunk 5]
```diff
@@ -53,19 +60,100 @@ impl Lexer {
...
+
     fn next_kind(&mut self) -> Result<Kind> {
         if self.start_of_line {
             if let Some(indent_kind) = self.match_indentation() {
                 self.start_of_line = false; // WHY!?
                 return Ok(indent_kind);
             }
         }
+        if self.fstring_stack.len() > 0 {
+            println!("going to match fstring token");
```
Comment (frustak) [inline src/lexer/lexer.rs:152]: hmm

File: src/lexer/lexer.rs [hunk 6]
```diff
@@ -404,6 +495,10 @@ impl Lexer {
         self.source[self.current..].chars().nth(1)
     }
 
+    fn triple_peek(&self) -> Option<char> {
+        self.source[self.current..].chars().nth(2)
+    }
```
Comment (frustak) [inline src/lexer/lexer.rs:500]: In future maybe we can add a method which takes peek number

File: src/lexer/lexer.rs [hunk 7]
```diff
@@ -727,6 +835,7 @@ mod tests {
                 if token.kind == Kind::Eof {
                     break;
                 }
+                println!("{:?}", token);
```
Comment (frustak) [inline src/lexer/lexer.rs:838]: nooo

File: src/lexer/lexer.rs [hunk 8]
```diff
@@ -923,20 +1018,20 @@ mod tests {
...
+        //         "rf\"hello\"",
+        //         "rf\"world\"",
+        //         "rf\"\"",
+        //         "a = rf\"hello\"",
+        //         "rf'hello_{var}'",
+        //         "rf\"\"\"hello\"\"\"",
+        //         "rf'''hello'''",
+        //     ],
+        // )
+        // .unwrap();
```
Comment (frustak) [inline src/lexer/lexer.rs:1034]: Why comment?

**Review by frustak — APPROVED** [review]
YAY, this is PR #100

File: src/lexer/diagnostics.rs (hunk 1 above)
Comment (Glyphack) [reply src/lexer/diagnostics.rs:5]: Prints out the given string to this error. Makes it possible to write

Syntax error err_str

File: src/lexer/lexer.rs (hunk 3 above)
Comment (Glyphack) [reply src/lexer/lexer.rs:73]: Peek token gets the next token but it saves the current state to rewind it after getting the token. This makes it possible for the parser to know what token is going to come after the current token without actually consuming that token.

This is useful for the parser to decide which rule is going to match the current expression.

I think the peek means looking which implies that this is just looking at the next token not consuming it

File: src/lexer/lexer.rs (hunk 6 above)
Comment (Glyphack) [reply src/lexer/lexer.rs:500]: I think this is the max peek we might need based on the python tokens present now. So probably we won't need another peek_4

**Pull request merged**